text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,299 @@
+
+# Prior Art
+[prior-art]: #prior-art
+
+- The term "non-idempotent" is used in section 9.1.2 of [Hypertext Transfer Protocol -- HTTP/1.1 (RFC 2616)](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Do we want (or need) a `CNB_STACK_TYPE` env var so that a stack buildpack can behave differently on each part of the stack?
+- Should the stackpack's detect have read-only access to the app?
+    - This would likely be driven by a stackpack that does not provide mixins, but instead dynamically contributes to the build plan based on the contents of the app source code. I don't know if we have a use case for this, but I can imaging a buildpack that reads environment variables as input to some function.
+- Should stackpacks be able to define per-stack mixins?
+    - We could support a top-level `mixins` array, and allow refinements under `[[stacks]] mixins`. If we do this, we need to distinguish between provided and required mixins (at least under `[[stacks]]`).
+    - If userspace buildpacks can require mixins from `bin/detect`, the stackpack could use this mechanism to require per-stack mixins.
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+A number of changes to the Platform Specification will be required to execute Stack Buildpacks. Those changes will be defined in a separate RFC.
+
+## Stack buildpacks
+
+Stack buildpacks are identical to other buildpacks, with the following exceptions:
+
+1. The `<layers>` directory is NOT writable.
+1. The working directory WILL NOT contain application source code during the build phase.
+1. All changes made to the filesystem during the execution of the stackpack's `bin/build` will be snapshotted and stored as a single layer, with the exception of the following directories:
+
+* `/tmp`
+* `/cnb`
+* `/layers`
+* `/workspace`
+* `/dev`
+* `/sys`
+* `/proc`
+* `/var/run/secrets`
+* `/etc/hostname`, `/etc/hosts`, `/etc/mtab`, `/etc/resolv.conf`
+* `/.dockerenv`
+
+## launch.toml (TOML)
+
+```
+[[excludes]]
+paths = ["<sub-path glob>"]
+cache = false
+```
+
+Where:
+
+* `paths` = a list of paths to exclude from the layer
+* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.
+
+## buildpack.toml  (TOML)
+
+ This proposal adds new keys to the `[buildpack]` table in `buildpack.toml`, and a new `[[mixins]]` array of tables:
+
+ ```
+[buildpack]
+privileged = <boolean (default=false)>
+non-idempotent = <boolean (default=false)>
+
+[buildpack.mixins]
+any = "<boolean (default=false)>"
+names = [ "<mixin name>" ]
+ ```
+
+* `privileged` - when set to `true`, the lifecycle will run this buildpack as the `root` user.
```

Comment:
Yes, I've left it somewhat generic because in the future we could add a feature that would allow `root` access for buildpacks that are not stackpacks (but that would break `rebase`)